rotkehlchen/balances/historical.py

```diff
@@ -1,29 +1,187 @@
+            amount=deserialize_fval(
+                value=entry[5],
+                name='amount',
+                location='historical balances',
+            ),
+            rate=deserialize_fval(
+                value=entry[6],
+                name='rate',
+                location='historical balances',
+            ),
+        )
+
+    def update_balances(self, current_balances: dict[Asset, FVal]) -> str | None:
+        """Updates current balances for a trade event, checking for negative balances.
+
+        Returns the event_identifier if a negative balance would occur, else None.
+        For trades:
+        - BUY means we receive base_asset and spend quote_asset
+        - SELL means we spend base_asset and receive quote_asset
+        """
+        if self.trade_type in {TradeType.BUY, TradeType.SETTLEMENT_BUY}:
+            spend_asset = self.quote_asset
+            receive_asset = self.base_asset
+            spend_amount = self.amount * self.rate
+            receive_amount = self.amount
+        else:
+            spend_asset = self.base_asset
+            receive_asset = self.quote_asset
+            spend_amount = self.amount
+            receive_amount = self.amount * self.rate
+
+        if current_balances[spend_asset] - spend_amount < ZERO:
+            return self.identifier
+
+        current_balances[spend_asset] -= spend_amount
+        current_balances[receive_asset] += receive_amount
+        if current_balances[spend_asset] == ZERO:
+            del current_balances[spend_asset]
+
+        return None
+
+
+@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=True)
+class MinimalHistoryEvent:
+    """Minimal history event affecting balances. Used for balance calculations."""
+    timestamp: Timestamp
+    identifier: str
+    amount: FVal
+    asset: Asset
+    event_type: HistoryEventType
+    event_subtype: HistoryEventSubType
+
+    @classmethod
+    def deserialize(cls, entry: tuple[Any, ...]) -> 'MinimalHistoryEvent':
+        """
+        May raise:
+        - UnknownAsset if the asset in the event cannot be found in the DB.
+        - DeserializationError
+        """
+        return cls(
+            identifier=entry[1],
+            amount=deserialize_fval(
+                value=entry[3],
+                name='amount',
+                location='historical balances',
+            ),
+            asset=Asset(entry[2]).check_existence(),
+            timestamp=deserialize_timestamp(entry[0]),
+            event_type=HistoryEventType.deserialize(entry[4]),
+            event_subtype=HistoryEventSubType.deserialize(entry[5]),
+        )
+
+    @staticmethod
+    def _get_direction_from_event(
+            event_type: HistoryEventType,
+            event_subtype: HistoryEventSubType,
+    ) -> EventDirection | None:
+        """Determine the direction of a history event based on its type and subtype.
+
+        This was copied from rotkehlchen/history/events/structures/base.py
```

Comment:
Yep ... as I said. Hacky as fuck. you don't copy code like this. This alone should indicate that this approach is completely wrong.